File: tools/java/GenTags.java
import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.util.*;
import java.util.zip.*;

/**
 * GenTags — Extracts damage-type and item tag membership from the 26.2 server jar.
 *
 * Reads the datapack tag JSON directly from the inner jar zip (the SAME data vanilla
 * net.minecraft.tags.TagLoader reads from the resource pack) — no Bootstrap binding,
 * no reflective mid-tick extraction. The datapack JSON is the authoritative source of
 * truth (CONTEXT Grey Area 1: "datapack/registries report is authoritative").
 *
 * Source semantics ported 1:1 from net.minecraft.tags.TagLoader.build:
 *   - A tag file is { "values": [ <entry>, ... ] }.
 *   - An <entry> is either a member resource-location string ("minecraft:fall"),
 *     or a REFERENCE to another tag, prefixed with '#' ("#minecraft:is_player_attack").
 *     A '#'-entry is RECURSIVELY expanded and its members merged (NOT stored literally) —
 *     exactly TagLoader.build's reference resolution.
 *   - An <entry> may also be the object form { "id": "...", "required": <bool> }
 *     (datapack value schema); the 26.2 jar uses only the bare-string form, but the
 *     object form is handled for forward-compatibility.
 *
 * Output: tags.json in the current directory (the extractor cwd = outputDir), shape:
 *   {
 *     "damage_type": { "bypasses_armor": ["minecraft:fall", ...], ... },
 *     "item":        { "pig_food": ["minecraft:carrot", ...], ... },
 *     "damage_type_elements": ["minecraft:arrow", "minecraft:fall", ...]   (sorted; the
 *         full damage_type registry element set, which has NO numeric id in registries.json
 *         because damage_type is a dynamic/datapack registry — the Go generator assigns ids)
 *   }
 * Member values are emitted as their FLATTENED resource-location strings; the Go generator
 * (tools/gen_tags.go) resolves them to numeric ids.
 */
public class GenTags {

    static final String DAMAGE_TYPE_PREFIX = "data/minecraft/tags/damage_type/";
    static final String ITEM_PREFIX = "data/minecraft/tags/item/";
    static final String DAMAGE_TYPE_ELEMENT_PREFIX = "data/minecraft/damage_type/";

    static final Gson GSON = new Gson();

    public static void main(String[] args) throws Exception {
        File innerJar = findInnerJar(args);
        if (innerJar == null) {
            System.err.println("GenTags: could not locate the inner server jar (-inner.jar) on the classpath or args");
            System.exit(1);
        }
        System.out.printf("GenTags: reading tags from %s%n", innerJar);

        // Raw per-family entry lists (values verbatim, including '#'-refs), keyed by tag name.
        Map<String, List<String>> damageRaw = new TreeMap<>();
        Map<String, List<String>> itemRaw = new TreeMap<>();
        Set<String> damageElements = new TreeSet<>();

        try (ZipFile zip = new ZipFile(innerJar)) {
            for (var entries = zip.entries(); entries.hasMoreElements(); ) {
                ZipEntry e = entries.nextElement();
                String name = e.getName();
                if (e.isDirectory()) continue;

                if (name.startsWith(DAMAGE_TYPE_PREFIX) && name.endsWith(".json")) {
                    damageRaw.put(tagName(name, DAMAGE_TYPE_PREFIX), readValues(zip, e));
                } else if (name.startsWith(ITEM_PREFIX) && name.endsWith(".json")) {
                    itemRaw.put(tagName(name, ITEM_PREFIX), readValues(zip, e));
                } else if (name.startsWith(DAMAGE_TYPE_ELEMENT_PREFIX) && name.endsWith(".json")
                        && !name.startsWith(DAMAGE_TYPE_PREFIX)) {
                    // A damage_type registry element (e.g. data/minecraft/damage_type/fall.json).
                    String id = name.substring(DAMAGE_TYPE_ELEMENT_PREFIX.length(), name.length() - ".json".length());
                    damageElements.add("minecraft:" + id);
                }
            }
        }

        // Recursively flatten '#'-refs (TagLoader.build) into flat member sets, per family.
        Map<String, List<String>> damageFlat = flattenAll(damageRaw);
        Map<String, List<String>> itemFlat = flattenAll(itemRaw);

        writeJson(damageFlat, itemFlat, new ArrayList<>(damageElements));

        System.out.printf("GenTags: wrote tags.json (%d damage_type tags, %d item tags, %d damage_type elements)%n",
            damageFlat.size(), itemFlat.size(), damageElements.size());
    }

    /** tagName turns "data/minecraft/tags/damage_type/bypasses_armor.json" into "bypasses_armor". */
    static String tagName(String entryName, String prefix) {
        String rest = entryName.substring(prefix.length());
        return rest.substring(0, rest.length() - ".json".length());
    }

    /** readValues parses a tag file's "values" array into the verbatim entry strings. */
    static List<String> readValues(ZipFile zip, ZipEntry e) throws IOException {
        List<String> out = new ArrayList<>();
        try (InputStreamReader r = new InputStreamReader(zip.getInputStream(e), StandardCharsets.UTF_8)) {
            JsonObject obj = GSON.fromJson(r, JsonObject.class);
            if (obj == null || !obj.has("values")) return out;
            JsonArray arr = obj.getAsJsonArray("values");
            for (JsonElement el : arr) {
                if (el.isJsonPrimitive()) {
                    out.add(el.getAsString());
                } else if (el.isJsonObject()) {
                    // Object form: { "id": "...", "required": <bool> } — take the id verbatim
                    // (the '#' nested-ref marker, if any, is carried inside the id string).
                    JsonObject vo = el.getAsJsonObject();
                    if (vo.has("id")) out.add(vo.get("id").getAsString());
                }
            }
        }
        return out;
    }

    /** flattenAll resolves '#'-refs for every tag in a family (TagLoader.build semantics). */
    static Map<String, List<String>> flattenAll(Map<String, List<String>> raw) {
        Map<String, List<String>> flat = new TreeMap<>();
        for (String tag : raw.keySet()) {
            LinkedHashSet<String> members = new LinkedHashSet<>();
            resolve(tag, raw, members, new HashSet<>());
            List<String> sorted = new ArrayList<>(members);
            Collections.sort(sorted);
            flat.put(tag, sorted);
        }
        return flat;
    }

    /**
     * resolve recursively expands a tag's entries into member resource-locations, following
     * '#'-prefixed references to other tags in the SAME family (TagLoader.build). The seen
     * set guards against cyclic tag references (vanilla TagLoader also detects cycles).
     */
    static void resolve(String tag, Map<String, List<String>> raw, Set<String> members, Set<String> seen) {
        if (!seen.add(tag)) return; // cycle guard
        List<String> entries = raw.get(tag);
        if (entries == null) return;
        for (String entry : entries) {
            if (entry.startsWith("#")) {
                // Reference to another tag: strip '#' and the "minecraft:" namespace, recurse.
                String refTag = stripNamespace(entry.substring(1));
                resolve(refTag, raw, members, seen);
            } else {
                members.add(entry); // a member resource-location ("minecraft:fall")
            }
        }
    }

    /** stripNamespace removes the "minecraft:" namespace from a tag reference. */
    static String stripNamespace(String s) {
        int i = s.indexOf(':');
        return i >= 0 ? s.substring(i + 1) : s;
    }

    /** findInnerJar locates the inner server jar via args[0] or the java.class.path. */
    static File findInnerJar(String[] args) {
        if (args.length > 0) {
            File f = new File(args[0]);
            if (f.exists()) return f;
        }
        // ExtractAll puts the inner jar FIRST on the classpath (buildExtractorClasspath).
        String cp = System.getProperty("java.class.path", "");
        for (String part : cp.split(File.pathSeparator)) {
            if (part.endsWith("-inner.jar")) {
                File f = new File(part);
                if (f.exists()) return f;
            }
        }
        return null;
    }

    /** writeJson hand-writes tags.json (PrintWriter + jsonStr, no external JSON emit dep). */
    static void writeJson(Map<String, List<String>> damage, Map<String, List<String>> item,
                          List<String> damageElements) throws IOException {
        try (PrintWriter pw = new PrintWriter(new FileWriter("tags.json"))) {
            pw.println("{");
            pw.print("  \"damage_type\": ");
            writeTagMap(pw, damage, "  ");
            pw.println(",");
            pw.print("  \"item\": ");
            writeTagMap(pw, item, "  ");
            pw.println(",");
            pw.print("  \"damage_type_elements\": ");
            writeStringArray(pw, damageElements, "  ");
            pw.println();
            pw.println("}");
        }
    }

    static void writeTagMap(PrintWriter pw, Map<String, List<String>> m, String indent) {
        pw.println("{");
        List<String> keys = new ArrayList<>(m.keySet());
        for (int i = 0; i < keys.size(); i++) {
            String k = keys.get(i);
            pw.printf("%s  %s: ", indent, jsonStr(k));
            writeStringArrayInline(pw, m.get(k));
            pw.println(i < keys.size() - 1 ? "," : "");
        }
        pw.printf("%s}", indent);
    }

    static void writeStringArray(PrintWriter pw, List<String> a, String indent) {
        pw.print("[");
        for (int i = 0; i < a.size(); i++) {
            pw.print(jsonStr(a.get(i)));
            if (i < a.size() - 1) pw.print(", ");
        }
        pw.print("]");
    }

    static void writeStringArrayInline(PrintWriter pw, List<String> a) {
        pw.print("[");
        for (int i = 0; i < a.size(); i++) {
            pw.print(jsonStr(a.get(i)));
            if (i < a.size() - 1) pw.print(", ");
        }
        pw.print("]");
    }

    static String jsonStr(String s) {
        return "\"" + s.replace("\\", "\\\\").replace("\"", "\\\"") + "\"";
    }
}

File: tools/gen_tags.go
// gen_tags generates data/tag/tags.go from tags.json (the GenTags.java extractor output).
//
// tags.json carries the recursively-flattened damage-type and item tag membership (GenTags.java
// resolves nested '#'-refs via net.minecraft.tags.TagLoader.build semantics before emitting it),
// plus the full damage_type registry element set.
//
// This generator emits the data/tag package read by the mob damage pipeline so
// damageSource.is("bypasses_armor") / .is("panic_causes") are GENUINE jar-data reads, NOT
// const-false stubs (PITFALLS Pitfall 3 — the keystone deliverable for MOB-SUB-03).
//
// Id resolution:
//   - Item member ids resolve through registries.json's minecraft:item registry (protocol_id ==
//     the slot index, exactly as gen_registryid builds data/registryid/item.go). An item member
//     that is absent from the registry is skipped (defensive; the jar's item tags only reference
//     real items).
//   - Damage-type ids have NO entry in registries.json — minecraft:damage_type is a DYNAMIC
//     (datapack) registry, so its protocol ids are assigned at server load, not in the static
//     report. We therefore assign each damage-type element a deterministic id = its index in the
//     sorted element list (GenTags emits damage_type_elements sorted). The id is internal server
//     state (DamageSource is constructed server-side; the damage-type id is never a wire registry
//     id), so a stable generated ordering is faithful AND reproducible. The generated DamageTypeIDs
//     map + DamageTypeNames slice expose the name<->id mapping so consumers (and tests) never hard-
//     code magic numbers.
//
// FRONT-LOAD NOTE (CONTEXT Grey Area 1): ItemTags is GENERATED here but has NO consumer in
// Phase 29 — the consumer is Phase 32 (TemptGoal). This is intentional shared-tooling front-
// loading: the data is real and tested headless (data/tag/tags_test.go), NOT a built-but-unwired
// violation.
//
// BLOCK-TAG NOTE: the current GenTags.java extractor emits ONLY damage_type + item tags into
// tags.json -- it does NOT yet emit block tags. The BlockTags table in data/tag/tags.go (the
// recursively-flattened mineable/pickaxe|axe|shovel|hoe membership that the Tool component mining
// rules read via BlockState.is) is therefore maintained OUT-OF-BAND: extracted directly from the
// jar's data/minecraft/tags/block/mineable/*.json (recursively resolving nested '#'-refs, the same
// TagLoader.build semantics GenTags applies to item/damage_type). Until GenTags is extended to add
// a "block" section to tags.json, a plain regen of this file does NOT reproduce BlockTags -- the
// block-tag block must be re-appended from the jar. This is a documented follow-up, not a silent
// gap: the data is real, jar-derived, and tested (data/tag/tags_test.go).
package main

import (
	"fmt"
	"go/format"
	"path/filepath"
	"sort"
	"strings"
)

// tagsReport is the GenTags.java output shape (tags.json).
type tagsReport struct {
	DamageType          map[string][]string `json:"damage_type"`
	Item                map[string][]string `json:"item"`
	DamageTypeElements  []string            `json:"damage_type_elements"`
}

// genTags reads tags.json + registries.json and emits data/tag/tags.go.
func genTags(jsonDir, goMCRoot string) error {
	jsonPath := filepath.Join(jsonDir, "tags.json")
	out := filepath.Join(goMCRoot, "data", "tag", "tags.go")

	var report tagsReport
	if err := readJSON(jsonPath, &report); err != nil {
		return fmt.Errorf("genTags: %w", err)
	}

	// --- damage-type id assignment (sorted element index) -----------------------------------
	dtElements := append([]string(nil), report.DamageTypeElements...)
	sort.Strings(dtElements)
	dtID := make(map[string]int32, len(dtElements)) // "minecraft:fall" -> id
	for i, name := range dtElements {
		dtID[name] = int32(i)
	}

	// --- item id resolution via registries.json (protocol_id == slot index) -----------------
	var registries registriesJSON
	if err := readJSON(filepath.Join(jsonDir, "registries.json"), &registries); err != nil {
		return fmt.Errorf("genTags: reading registries.json: %w", err)
	}
	itemReg, ok := registries["minecraft:item"]
	if !ok {
		return fmt.Errorf("genTags: registries.json has no minecraft:item registry")
	}
	itemID := make(map[string]int32, len(itemReg.Entries)) // "minecraft:carrot" -> protocol id
	for name, v := range itemReg.Entries {
		itemID[name] = int32(v.ProtocolID)
	}

	// --- build sorted membership rows for deterministic output ------------------------------
	damageRows := buildTagRows(report.DamageType, dtID)
	itemRows := buildTagRows(report.Item, itemID)

	var buf strings.Builder
	buf.WriteString(generatedHeader("gen_tags.go", "tags.json", "registries.json"))
	buf.WriteByte('\n')
	buf.WriteString("package tag\n\n")

	buf.WriteString(`// DamageTypeNames maps a generated damage-type id back to its registry resource id
// ("minecraft:fall"). The id is the element's index in the sorted damage_type registry element
// set — a deterministic, internal-only ordering (minecraft:damage_type is a dynamic registry with
// no static protocol id; the id is never a wire registry id, only server-side DamageSource state).
var DamageTypeNames = []string{
`)
	for _, name := range dtElements {
		fmt.Fprintf(&buf, "\t%q,\n", name)
	}
	buf.WriteString("}\n\n")

	buf.WriteString(`// DamageTypeIDs maps a damage-type registry resource id ("minecraft:fall") to its generated id.
// Consumers resolve a DamageSource's type by name; tests assert membership by name (no magic ids).
var DamageTypeIDs = map[string]int32{
`)
	for _, name := range dtElements {
		fmt.Fprintf(&buf, "\t%q: %d,\n", name, dtID[name])
	}
	buf.WriteString("}\n\n")

	buf.WriteString(`// DamageTypeTags maps a damage-type tag name (e.g. "bypasses_armor") to the SET of damage-type
// ids in it, recursively flattened (nested #-tag refs resolved at extraction time, like vanilla
// net.minecraft.tags.TagLoader.build). damageSource.is(tag) is DamageTypeTags[tag][typeID].
var DamageTypeTags = map[string]map[int32]bool{
`)
	writeTagSets(&buf, damageRows)
	buf.WriteString("}\n\n")

	buf.WriteString(`// ItemTags maps an item tag name (e.g. "pig_food") to the SET of item ids in it (recursively
// flattened, same as DamageTypeTags). GENERATED-BUT-NOT-YET-WIRED in Phase 29 — the consumer is
// Phase 32 (TemptGoal). Intentional shared-tooling front-load (CONTEXT Grey Area 1); the data is
// real and tested headless, not a built-but-unwired violation.
var ItemTags = map[string]map[int32]bool{
`)
	writeTagSets(&buf, itemRows)
	buf.WriteString("}\n")

	formatted, err := format.Source([]byte(buf.String()))
	if err != nil {
		return fmt.Errorf("genTags: gofmt: %w", err)
	}
	if err := writeFile(out, formatted); err != nil {
		return fmt.Errorf("genTags: %w", err)
	}
	logf("genTags: wrote %s (%d damage_type tags, %d item tags, %d damage_type elements)",
		out, len(damageRows), len(itemRows), len(dtElements))
	return nil
}

// tagSetRow is one tag's resolved member id set.
type tagSetRow struct {
	Name string
	IDs  []int32
}

// buildTagRows resolves each tag's member resource-locations to numeric ids via idOf, dropping
// members that don't resolve (defensive), and returns rows sorted by tag name with sorted ids.
func buildTagRows(tags map[string][]string, idOf map[string]int32) []tagSetRow {
	rows := make([]tagSetRow, 0, len(tags))
	for name, members := range tags {
		seen := make(map[int32]bool, len(members))
		ids := make([]int32, 0, len(members))
		for _, m := range members {
			id, ok := idOf[m]
			if !ok {
				continue // unresolved member (not in the registry) — skip
			}
			if seen[id] {
				continue
			}
			seen[id] = true
			ids = append(ids, id)
		}
		sort.Slice(ids, func(i, j int) bool { return ids[i] < ids[j] })
		rows = append(rows, tagSetRow{Name: name, IDs: ids})
	}
	sort.Slice(rows, func(i, j int) bool { return rows[i].Name < rows[j].Name })
	return rows
}

// writeTagSets emits the "name": {id: true, ...} body for a map[string]map[int32]bool literal.
func writeTagSets(buf *strings.Builder, rows []tagSetRow) {
	for _, r := range rows {
		fmt.Fprintf(buf, "\t%q: {", r.Name)
		for i, id := range r.IDs {
			if i > 0 {
				buf.WriteString(", ")
			}
			fmt.Fprintf(buf, "%d: true", id)
		}
		buf.WriteString("},\n")
	}
}
